src/tensor/cache.rs

```diff
@@ -1,24 +1,70 @@
-use std::{alloc::Layout, collections::BTreeMap, vec::Vec};
+use std::{
+    alloc::Layout,
+    collections::{BTreeMap, VecDeque},
+    vec::Vec,
+};
+
+#[cfg(not(feature = "no-std"))]
+use std::vec;
+
+#[cfg(feature = "no-std")]
+use alloc::vec;
 
 #[cfg(not(feature = "no-std"))]
 use std::sync::RwLock;
 
 #[cfg(feature = "no-std")]
 use spin::RwLock;
 
-/// A key for the tensor cache. Contains both number of bytes and informatino
+macro_rules! read {
+    ($x:expr) => {{
+        #[cfg(not(feature = "no-std"))]
+        {
+            $x.read().unwrap()
+        }
+        #[cfg(feature = "no-std")]
+        {
+            $x.read()
+        }
+    }};
+}
+
+macro_rules! write {
+    ($x:expr) => {{
+        #[cfg(not(feature = "no-std"))]
+        {
+            $x.write().unwrap()
+        }
+        #[cfg(feature = "no-std")]
+        {
+            $x.write()
+        }
+    }};
+}
+
+/// A key for the tensor cache. Contains both number of bytes and information
 /// about the layout of the allocation.
 ///
 /// Since [Layout] doesn't impl Ord, we can't use it directly as a key
-/// for a hasmap, meaning we need this extra datastructure. Otherwise
+/// for a hashmap, meaning we need this extra datastructure. Otherwise
 /// we could just using `(usize, Layout)` as the key.
 #[derive(Debug, Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub(crate) struct AllocationKey {
-    pub num_bytes: usize,
-    /// The size of the allocation in bytes - from [Layout].
-    pub size: usize,
+struct AllocationKey {
+    /// The size of the allocation in bytes
+    num_bytes: usize,
+    /// The size of the type in bytes - from [Layout].
+    size: usize,
     /// The alignment of the allocation in bytes - from [Layout].
-    pub alignment: usize,
+    alignment: usize,
+}
+
+#[derive(Debug)]
+struct AllocationGroup<Ptr: CacheStorage> {
+    // Tracks the number of matching 'AllocationKey's in drop_queue to ignore. This is used to
+    // "remove" the next instance of the matching AllocationKey in the drop_queue, without having
+    // to run an O(n)
```

Comment:
I think I'm missing something - an O(n) what? `vec.remove(0)` would have O(n) performance, but `vec.pop()` shouldn't do any moving of the data.

Reply:
This should be better explained now.